cores/esp8266/erase_config.cpp

```diff
@@ -0,0 +1,461 @@
+   PLL adjustments are preserved, the UART date rate is 115200 bps.
+
+   The solution impelmented here is to adjust the UART data rate to  an
+   effective rate of 74880 based on the kind of reset. This minimizes the
+   garbled printing.
+
+   Assumes a data rate of 74880.
+   miniterm.py does a good job of handling this rate on Linux.
+ */
+
+/* ROM reset reason values returned by `rtc_get_reset_reason` are different from
+   the NONOS SDK reset reasons. */
+typedef enum ROM_RST_REASON { /* Comments on the right are from RTOS SDK */
+    NO_MEAN                =  0,    /* Undefined */
+    POWERON_RESET          =  1,    /* Power on boot      *//**<1, Vbat power on reset */
+    EXT_RESET              =  2,    /* External reset or wake-up from Deep-sleep */
+                                                            /**<2, external system reset */
+    SW_RESET               =  3,    /*                    *//**<3, Software reset digital core */
+    OWDT_RESET             =  4,    /* Hardware WDT reset *//**<4, Legacy watch dog reset digital core */
+    DEEPSLEEP_RESET        =  5,    /*                    *//**<5, Deep Sleep reset digital core */
+    SDIO_RESET             =  6,    /*                    *//**<6, Reset by SLC module, reset digital core*/
+} ROM_RST_REASON_t;
+
+#define RTC_SYS ((volatile uint32_t*)0x60001100)
+extern uint32_t rtc_get_reset_reason(void);
+
+bool IRAM_ATTR is_cpu_freq_cal(void) {
+    const uint32_t rtc_sys_reason = RTC_SYS[0];
+    const uint32_t rom_api_reason = rtc_get_reset_reason();
+    if (1 >= rtc_sys_reason && OWDT_RESET != rom_api_reason) {
+        return false;  // REASON_EXT_SYS_RST
+    }
+    if (REASON_EXT_SYS_RST < rtc_sys_reason) {
+        return false;  // REASON_DEFAULT_RST
+    }
+    return true;
+}
+
+void IRAM_ATTR erase_config__fix_divider(void) {
+    /*
+       When reset cause is not power-on or EXT_RST the CPU crystal calibration
+       has been done and there is no need to correct. With the exception of some
+       early callback(s) from the SDK like run_user_rf_pre_init(). The UART
+       speed appears to always be based on the precalibrated crystal frequency.
+     */
+    uint32_t divider = UART_CLK_FREQ / 115200;
+    if (is_cpu_freq_cal()) {
+        divider = UART_CLK_FREQ / 74880;
+    }
+    /*
+      Here we use uart_div_modify in the Boot ROM. Note the Boot ROM version does
+      not do any input validation.
+
+      The SDK has an overide on uart_div_modify. We cannot use its replacement. It
+      is not in IRAM and reauires SDK initialization.
+     */
+    rom_uart_div_modify(0, divider);
+    ERASE_CFG__ETS_DELAY_US(150);
+}
+
+/*
+  Something to see when we are in the SDK initialization.
+*/
+extern "C" void IRAM_ATTR _Z22__run_user_rf_pre_initv(void) {
+    real_uart_div_modify(0, UART_CLK_FREQ / 115200);
+    ERASE_CFG__ETS_DELAY_US(150);
+    ERASE_CFG__ETS_PRINTF("\n__run_user_rf_pre_init()\n");
+    ERASE_CFG__ETS_FLUSH(0);
+}
+
+#if 1
```

Comment:
constant ifdef

Reply:
Also, I'm not sure I get why you're overloading pinMode as I don't see it used anywhere in this file.

Reply:
Debug code to preserve UART operation during the SDK/Core startup so a few more debug messages could print out. Without it, the UART was killed and some messages garbled.